### PR TITLE
Add 'InputDevice' abstraction and 'Camera' constr.

### LINKED ISSUE #15: Support for access to input devices (libavdevice)
I have patched ffmpeg to support input devices, such as avfoundation.  If I bother to polish it would you accept a pull? It would probably result in changing imageReader to something like:

```
data Input = File FIlePath | Camera String
imageReader :: Input -> IO (IO PictureThing, IO ())
```

### PR DESCRIPTION
**Pull request by TomMD**
Fix #15

### CODE REVIEW

**Comment by TomMD** [comment]
Notice CI fails due to an old version of libavdevice (it appears).

File: src/Codec/FFmpeg/Types.hsc [hunk 1]
```diff
@@ -24,12 +25,45 @@ newtype AVFormatContext = AVFormatContext (Ptr ()) deriving (Storable, HasPtr)
...
+#hasField AVFormatContext, InputFormat, iformat
 #hasField AVFormatContext, IOContext, pb
 
+setFilename :: AVFormatContext -> String -> IO ()
+setFilename ctx fn =
+    do let ptr  = getPtr ctx
+           dst   = (#ptr AVFormatContext, filename) ptr
+           bytes = map (fromIntegral . fromEnum) fn
+       zipWithM_ (pokeElemOff dst) bytes [(0 :: CInt) ..]
+
```
Comment (acowley) [inline src/Codec/FFmpeg/Types.hsc]: Should we instead use `withCString` here?

**Comment by acowley** [comment]
The Travis tests are only vaguely useful as the libraries are so old. They primarily serve to make sure it builds on not-my-machine (for any appropriate definition of "my").

This branch builds for me on OS X. I wrote a demo of the camera functionality, but I get an error that the `Selected framerate (29.970030) is not supported by the device`. I've made a PR to your branch so you can take a look. Once we get that working on two computers, we can rebase to clean up the history and merge into master.

**Comment by TomMD** [comment]
Hey, it's not just me!  In short: this bug is down in FFMPEG and I'll look into it more.

I did get that same framerate error but after hard-coding FFMPEG (the C library) to always use 30FPS right at the point of consumption the issue went away.  Indeed, even though the device and command line flag (when I provided it) says 30fps the float at said point is still 29.97.

**Comment by TomMD** [comment]
I merged your pull and, ignoring the 29.97 bug (using my hard-coded 30FPS work-around), I also get an issue with encoding with Mpeg4:

```
[mpeg4 @ 0x7fc011835400] framerate not set
Main: Couldn't open codec
```

Notice this is just the pulse demo - no camera required:

```
ffmpeg-light% ./demo/Main
[mpeg4 @ 0x7fc011835400] framerate not set
Main: Couldn't open codec
```

Do you have thoughts on this?  I haven't looked into it yet beyond seeing the C code checks if the  framerate (rational) has a zero numerator or denominator and fails.

**Comment by TomMD** [comment]
I'd like to close on this but am fighting with myself as to how.

1. One way, which I don't have time or motivation for, is to use FFMPEG prims to query the system and determine the correct settings (framerate in our current case) and apply those, since FFMPEG frustratingly decides to bail instead of use the available settings.

2. We can include another field in the `Cam` constructor that is optional settings, including framerate, which we pass to openCamera (and use in `dict`).  I like this method.

3. Forget it all, 30 fps is good enough for everyone, and merge.  Call me lazy, but I'm rather OK with this actually.

Thoughts?

Context: I'd like to make another version of my mellow library (kinect depth images -> friday image processing -> gloss) that uses visible spectrum/webcam as the input and this is the first step.

**Comment by acowley** [comment]
I'm fine with option 3 as the way to go until we figure out something better. Does this [newer upstream issue](https://trac.ffmpeg.org/ticket/4880) mean anything to you? I don't remember the details of this issue, tbh. So are we happy with the current state of things? `cabal run demo -- cam` works for me as things stand.

**Comment by TomMD** [comment]
I'm fine with things.

That bug report 1. Says to me there are people without 30 fps options so we
might see a follow up to this PR.  2. Had a common misuse if cli (-r
instead of -framerate).

**Comment by acowley** [comment]
Alright, let's keep the ball rolling. I'd like to try to get the travis builds working again, though. I know it's not hugely significant, but it's a nice safety blanket.

**Pull request merged**